llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,323 @@
+//===- llvm/ADT/PagedVector.h - 'Lazyly allocated' vectors --------*- C++
+//-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the PagedVector class.
+//
+//===----------------------------------------------------------------------===//
+#ifndef LLVM_ADT_PAGEDVECTOR_H
+#define LLVM_ADT_PAGEDVECTOR_H
+
+#include "llvm/ADT/PointerIntPair.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Allocator.h"
+#include <cassert>
+#include <vector>
+
+namespace llvm {
+/// A vector that allocates memory in pages.
+///
+/// Order is kept, but memory is allocated only when one element of the page is
+/// accessed. This introduces a level of indirection, but it is useful when you
+/// have a sparsely initialised vector where the full size is allocated upfront.
+///
+/// As a side effect the elements are initialised later than in a normal vector.
+/// On the first access to one of the elements of a given page all, the elements
+/// of the page are initialised. This also means that the elements of the page
+/// are initialised beyond the size of the vector.
+///
+/// Similarly on destruction the elements are destroyed only when the page is
+/// not needed anymore, delaying invoking the destructor of the elements.
+///
+/// Notice that this does not have iterators, because if you have iterators it
+/// probably means you are going to touch all the memory in any case, so better
+/// use a std::vector in the first place.
+template <typename T, size_t PageSize = 1024 / sizeof(T)> class PagedVector {
+  static_assert(PageSize > 1, "PageSize must be greater than 0. Most likely "
+                              "you want it to be greater than 16.");
+  /// The actual number of element in the vector which can be accessed.
+  size_t Size = 0;
+
+  /// The position of the initial element of the page in the Data vector.
+  /// Pages are allocated contiguously in the Data vector.
+  mutable SmallVector<T *, 0> PageToDataPtrs;
```

Comment:
Looks like there was a comment here: https://github.com/llvm/llvm-project/pull/66430#discussion_r1330540304

I guess the idea was to be able to merge the `InvalidPage` into this.

Reply:
Now InvalidPage is just holding nullptr. Do you mean you prefer I use nullptr directly?

Reply:
In fact I am fine either way, but I think that was what @zygoloid meant with his comment. However, the diff has changed quite a lot since then, maybe we should keep things as they are now.

Reply:
I would keep the special name, at least for now. To reduce memory usage, it might make sense to switch that back to be an integer index in some larger "slab", at which point one would keep a pointer to the "slab" and use -1 rather than 0 to indicate missing pages.

Reply:
I have no particular preference here. I don't think the switch back to an integer index is a likely future change (because that'd take us back to paying for a reallocation if the slab becomes full like with the original approach of using a vector), and it wouldn't be hard to change the places in this class that use `nullptr` as the invalid page over to using a different constant, so the evolution argument doesn't really persuade me. But having an explicit name for the invalid page makes the code marginally more readable.

Reply:
I will switch everything to nullptr and switch back.